Mathlib/Algebra/Category/CommAlg/Basic.lean

```diff
@@ -0,0 +1,310 @@
+    {g₁ : CommRing X₁} {g₂ : CommRing X₂} {m₁ : Algebra R X₁} {m₂ : Algebra R X₂}
+    (e : X₁ ≃ₐ[R] X₂) : CommAlg.of R X₁ ≅ CommAlg.of R X₂ where
+  hom := CommAlg.ofHom (e : X₁ →ₐ[R] X₂)
+  inv := CommAlg.ofHom (e.symm : X₂ →ₐ[R] X₁)
+
+namespace CategoryTheory.Iso
+
+/-- Build a `AlgEquiv` from an isomorphism in the category `CommAlg R`. -/
+@[simps]
+def commAlgIsoToAlgEquiv {X Y : CommAlg R} (i : X ≅ Y) : X ≃ₐ[R] Y where
+  __ := i.hom.hom
+  toFun := i.hom
+  invFun := i.inv
+  left_inv x := by simp
+  right_inv x := by simp
+
+end CategoryTheory.Iso
+
+/-- Algebra equivalences between `Algebra`s are the same as isomorphisms in `CommAlg`. -/
+@[simps!]
+def algEquivEquivCommAlgIso {X Y : Type u} [CommRing X] [CommRing Y] [Algebra R X] [Algebra R Y] :
+    (X ≃ₐ[R] Y) ≃ (CommAlg.of R X ≅ CommAlg.of R Y) where
+  toFun := AlgEquiv.toCommAlgIso
+  invFun := Iso.commAlgIsoToAlgEquiv
+  left_inv _ := rfl
+  right_inv _ := rfl
+
+instance CommAlg.forget_reflects_isos : (forget (CommAlg.{u} R)).ReflectsIsomorphisms where
+  reflects {X Y} f _ := by
+    let i := asIso ((forget (CommAlg.{u} R)).map f)
+    let e : X ≃ₐ[R] Y := { f.hom, i.toEquiv with }
+    exact e.toCommAlgIso.isIso_hom
+
+namespace CommAlg
+
+noncomputable section Coprod
+
+open TensorProduct
+
+variable (A B C : CommAlg R)
+
+/-- The explicit cocone with tensor products as the fibered product in `CommRingCat`. -/
+def binaryCofan : Limits.BinaryCofan A B :=
+  Limits.BinaryCofan.mk (ofHom Algebra.TensorProduct.includeLeft)
+    (ofHom (Algebra.TensorProduct.includeRight (R := R) (A := A)))
+
+@[simp]
+lemma binaryCofan_inl : (binaryCofan A B).inl = ofHom Algebra.TensorProduct.includeLeft := rfl
+
+@[simp]
+lemma binaryCofan_inr : (binaryCofan A B).inr = ofHom Algebra.TensorProduct.includeRight := rfl
+
+@[simp] lemma binaryCofan_pt : (binaryCofan A B).pt = .of R (A ⊗[R] B) := rfl
+
+/-- Verify that the `pushout_cocone` is indeed the colimit. -/
+def binaryCofanIsColimit : Limits.IsColimit (binaryCofan A B) :=
+  Limits.BinaryCofan.IsColimit.mk _
+    (fun f g ↦ ofHom (Algebra.TensorProduct.lift f.hom g.hom fun _ _ ↦ .all _ _))
+    (fun f g ↦ by ext1; exact Algebra.TensorProduct.lift_comp_includeLeft _ _ fun _ _ ↦ .all _ _)
+    (fun f g ↦ by ext1; exact Algebra.TensorProduct.lift_comp_includeRight _ _ fun _ _ ↦ .all _ _)
+    (fun f g m hm₁ hm₂ ↦ by
+      ext1
+      refine Algebra.TensorProduct.liftEquiv.symm_apply_eq (y := ⟨⟨_, _⟩, fun _ _ ↦ .all _ _⟩).mp ?_
+      exact Subtype.ext (Prod.ext congr(($hm₁).hom) congr(($hm₂).hom)))
+
+def isInitialSelf : Limits.IsInitial (of R R) := .ofUniqueHom (fun A ↦ ofHom (Algebra.ofId R A))
+  fun _ _ ↦ hom_ext (Algebra.ext_id _ _ _)
+
+open Opposite
+
+instance : ChosenFiniteProducts (CommAlg R)ᵒᵖ where
+  product A B := ⟨BinaryCofan.op <| (binaryCofan (unop A) (unop B)),
+    BinaryCofan.IsColimit.op <| (binaryCofanIsColimit (unop A) (unop B))⟩
+  terminal := ⟨_, terminalOpOfInitial isInitialSelf⟩
+
+open MonoidalCategory
```

Comment:
Can we we follow `Mathlib.Algebra.Category.AlgebraCat.Monoidal` here instead to get these as defeqs?

Reply:
Not sure how. `CommAlg` isn't merely monoidal, but cartesian-monoidal